From 0931a1af92a8ff076753b7d9993abaf25a616185 Mon Sep 17 00:00:00 2001
From: Przemyslaw Czarnota <przemyslawx.czarnota@intel.com>
Date: Wed, 6 Jun 2018 08:23:00 +0200
Subject: clsact: add clsact qdisc

clsact qdisc is needed to be able to classify egress traffic
---
 Makefile.am              |  1 +
 lib/route/qdisc/clsact.c | 64 ++++++++++++++++++++++++++++++++++++++++
 2 files changed, 65 insertions(+)
 create mode 100644 lib/route/qdisc/clsact.c

diff --git a/Makefile.am b/Makefile.am
index ab5481b..61e779f 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -439,6 +439,7 @@ lib_libnl_route_3_la_SOURCES = \
 	lib/route/qdisc/hfsc.c \
 	lib/route/qdisc/htb.c \
 	lib/route/qdisc/ingress.c \
+	lib/route/qdisc/clsact.c \
 	lib/route/qdisc/mqprio.c \
 	lib/route/qdisc/netem.c \
 	lib/route/qdisc/plug.c \
diff --git a/lib/route/qdisc/clsact.c b/lib/route/qdisc/clsact.c
new file mode 100644
index 0000000..6badc91
--- /dev/null
+++ b/lib/route/qdisc/clsact.c
@@ -0,0 +1,64 @@
+/*
+ * lib/route/qdisc/clsact.c		clsact
+ *
+ *	This library is free software; you can redistribute it and/or
+ *	modify it under the terms of the GNU Lesser General Public
+ *	License as published by the Free Software Foundation version 2.1
+ *	of the License.
+ *
+ * Copyright (c) 2018 Intel Corporation
+ */
+
+/**
+ * @ingroup qdisc
+ * @defgroup qdisc_clsact clsact qdisc
+ *
+ * @{
+ */
+
+#include <netlink-private/netlink.h>
+#include <netlink-private/tc.h>
+#include <netlink/netlink.h>
+#include <netlink-private/route/tc-api.h>
+#include <netlink/route/qdisc.h>
+#include <netlink/utils.h>
+
+struct dumb {
+	uint32_t foo;
+};
+
+static int dumb_msg_parser(struct rtnl_tc *tc, void *data)
+{
+	return 0;
+}
+
+static void dumb_dump_line(struct rtnl_tc *tc, void *data,
+			    struct nl_dump_params *p)
+{
+}
+
+static int dumb_msg_fill(struct rtnl_tc *tc, void *data, struct nl_msg *msg)
+{
+	return 0;
+}
+
+static struct rtnl_tc_ops clsact_ops = {
+	.to_kind		= "clsact",
+	.to_type		= RTNL_TC_TYPE_QDISC,
+	.to_size		= sizeof(struct dumb),
+	.to_msg_parser		= dumb_msg_parser,
+	.to_dump[NL_DUMP_LINE]	= dumb_dump_line,
+	.to_msg_fill		= dumb_msg_fill,
+};
+
+static void __init clsact_init(void)
+{
+	rtnl_tc_register(&clsact_ops);
+}
+
+static void __exit clsact_exit(void)
+{
+	rtnl_tc_unregister(&clsact_ops);
+}
+
+/** @} */
